### PR TITLE
Apr 1291 gateway trading name

### PR DESCRIPTION
**Pull request by MarkFCain**

### CODE REVIEW

File: src/SFA.DAS.AdminService.Web.Tests/Controllers/ControllerAuthorizeAttributeTests.cs [hunk 1]
```diff
@@ -15,7 +15,7 @@ public class ControllerAuthorizeAttributeTests
 
         private readonly List<string> _controllersThatDoNotRequireAuthorize = new List<string>()
         {
-            "AccountController", "ExcelAwareController", "HomeController", "ErrorPageController", "TestErrorController"
+            "AccountController", "ExcelAwareController", "HomeController", "ErrorPageController", "TestErrorController", "RoatpGatewayController"
```
Comment (martin-hornsby) [inline src/SFA.DAS.AdminService.Web.Tests/Controllers/ControllerAuthorizeAttributeTests.cs]: RoatpGatewayController requires Authorize as it's for RoATP only

File: src/SFA.DAS.AdminService.Web.Tests/Controllers/Gateway/RoatpGatewayControllerLegalNameTests.cs [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+            {
+                new Claim(ClaimTypes.NameIdentifier, "1"),
+                new Claim("http://schemas.xmlsoap.org/ws/2005/05/identity/claims/upn", username),
+                new Claim(ClaimTypes.GivenName, givenName),
+                new Claim(ClaimTypes.Surname, surname)
+            }));
+
+            var context = new DefaultHttpContext{ User = user };
+
+            _contextAccessor.Setup(_ => _.HttpContext).Returns(context);
```
Comment (martin-hornsby) [inline src/SFA.DAS.AdminService.Web.Tests/Controllers/Gateway/RoatpGatewayControllerLegalNameTests.cs]: Could we use the MockedHttpContextAccessor helper class from Assessor Service ?

File: src/SFA.DAS.AdminService.Web/Controllers/Roatp/Apply/RoatpGatewayController.cs [hunk 3]
```diff
@@ -172,11 +174,34 @@ public async Task<IActionResult> EvaluateLegalNamePage(LegalNamePageViewModel vi
             return RedirectToAction("ViewApplication", new { viewModel.ApplicationId });
         }
 
-        [HttpGet("/Roatp/Gateway/{applicationId}/Page/1-10")]
-        public async Task<IActionResult> GetGatewayLegalNamePage(Guid applicationId, string pageId)
+        [HttpGet("/Roatp/Gateway/{applicationId}/Page/1-20")]
+        public async Task<IActionResult> GetGatewayTradingNamePage(Guid applicationId, string pageId)
```
Comment (martin-hornsby) [inline src/SFA.DAS.AdminService.Web/Controllers/Roatp/Apply/RoatpGatewayController.cs]: Where's pageId coming from?  It needs to be decorated with [FromBody] if not coming from the URL Route

File: src/SFA.DAS.AdminService.Web/Handlers/Gateway/GetTradingNameHandler.cs [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+        private readonly IRoatpApplicationApiClient _applyApiClient;
+        private readonly IQnaApiClient _qnaApiClient;
+        private readonly IRoatpApiClient _roatpApiClient;
+
+
+        public GetTradingNameHandler(IRoatpApplicationApiClient applyApiClient, IQnaApiClient qnaApiClient, IRoatpApiClient roatpApiClient)
+        {
+            _applyApiClient = applyApiClient;
+            _qnaApiClient = qnaApiClient;
+            _roatpApiClient = roatpApiClient;
```
Comment (martin-hornsby) [inline src/SFA.DAS.AdminService.Web/Handlers/Gateway/GetTradingNameHandler.cs]: Do these work as expected if Dependency Injection is set to Transient?

File: src/SFA.DAS.AdminService.Web/Handlers/Gateway/GetTradingNameHandler.cs [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+
+        public async Task<TradingNamePageViewModel> Handle(GetTradingNameRequest request, CancellationToken cancellationToken)
+
+        {
+            var pageId = GatewayPageIds.TradingName;
+
+            var model = new TradingNamePageViewModel { ApplicationId = request.ApplicationId, PageId = pageId };
+
+            var currentRecord = await _applyApiClient.GetGatewayPageAnswer(request.ApplicationId, pageId);
```
Comment (martin-hornsby) [inline src/SFA.DAS.AdminService.Web/Handlers/Gateway/GetTradingNameHandler.cs]: assumption that GetGatewayPageAnswer is always successful

File: src/SFA.DAS.AdminService.Web/Handlers/Gateway/GetTradingNameHandler.cs [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+            var applicationDetails = await _applyApiClient.GetApplication(model.ApplicationId);
```
Comment (martin-hornsby) [inline src/SFA.DAS.AdminService.Web/Handlers/Gateway/GetTradingNameHandler.cs]: assumption that GetApplication is always successful

File: src/SFA.DAS.AdminService.Web/Handlers/Gateway/GetTradingNameHandler.cs [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+            {
+                gatewayReviewStatus = applicationDetails.GatewayReviewStatus;
+            }
+
+            if (currentRecord?.GatewayPageData != null)
+            {
+                model = JsonConvert.DeserializeObject<TradingNamePageViewModel>(currentRecord.GatewayPageData);
+                model.Status = currentRecord.Status;
+                model.GatewayReviewStatus = gatewayReviewStatus;
+                return model;
```
Comment (martin-hornsby) [inline src/SFA.DAS.AdminService.Web/Handlers/Gateway/GetTradingNameHandler.cs]: can this be re-written so you're not returning half way in a method?

File: src/SFA.DAS.AdminService.Web/Handlers/Gateway/GetTradingNameHandler.cs [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+        {
+            _applyApiClient = applyApiClient;
+            _qnaApiClient = qnaApiClient;
+            _roatpApiClient = roatpApiClient;
+        }
+
+
+        public async Task<TradingNamePageViewModel> Handle(GetTradingNameRequest request, CancellationToken cancellationToken)
+
+        {
```
Comment (martin-hornsby) [inline src/SFA.DAS.AdminService.Web/Handlers/Gateway/GetTradingNameHandler.cs]: This is a long method. Could it be split into smaller functions?

File: src/SFA.DAS.AdminService.Web/Validators/Roatp/RoatpGatewayPageViewModelValidator.cs [hunk 9]
```diff
@@ -9,8 +9,14 @@
...
 namespace SFA.DAS.AdminService.Web.Validators.Roatp
 {
-    public class RoatpGatewayPageViewModelValidator : IRoatpGatewayPageViewModelValidator
+    public  class RoatpGatewayPageViewModelValidator : IRoatpGatewayPageViewModelValidator
     {
+
+        private  string SelectAnOutcome => "Select the outcome of this check";
+
+        private string FailDetailsRequired => "Enter the reasons this check was failed";
+        private string TooManyWords => "Your comments must be 150 words or less";
```
Comment (martin-hornsby) [inline src/SFA.DAS.AdminService.Web/Validators/Roatp/RoatpGatewayPageViewModelValidator.cs]: should be private const string, not lambda functions

File: src/SFA.DAS.AdminService.Web/Views/Roatp/Apply/Gateway/Pages/TradingName.cshtml [hunk 10]
```diff
@@ -0,0 +1,307 @@
...
+@model SFA.DAS.AdminService.Web.ViewModels.Roatp.Gateway.TradingNamePageViewModel
+
+@{
+    var heading = "Trading name check";
+    var caption = "Organisation checks";
+
+    ViewBag.Title = $"Gateway - {heading}";
+    Layout = "_Layout";
+
+    var formGroupError = Model.ErrorMessages != null && Model.ErrorMessages.Any() ? "govuk-form-group--error" : "";
```
Comment (martin-hornsby) [inline src/SFA.DAS.AdminService.Web/Views/Roatp/Apply/Gateway/Pages/TradingName.cshtml]: Bit of a code smell here... you should be using ModelState.IsValid which is the design pattern used in most other places

File: src/SFA.DAS.AdminService.Web/Views/Roatp/Apply/Gateway/Pages/TradingName.cshtml [hunk 11]
```diff
@@ -0,0 +1,307 @@
...
+        </li>
+        
+        <li class="govuk-breadcrumbs__list-item" aria-current="page">@heading</li>
+    </ol>
+</div>
+<main class="govuk-main-wrapper " id="main-content" role="main">
+
+    <div class="govuk-grid-row">
+        <div class="govuk-grid-column-two-thirds">
+            @if (Model.ErrorMessages!=null && Model.ErrorMessages.Any())
```
Comment (martin-hornsby) [inline src/SFA.DAS.AdminService.Web/Views/Roatp/Apply/Gateway/Pages/TradingName.cshtml]: As mentiond. Should be using ModelState.isValid as per other views

File: src/SFA.DAS.AdminService.Web/Handlers/Gateway/GetTradingNameHandler.cs [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+            }
+
+            model.GatewayReviewStatus = gatewayReviewStatus;
+            if (applicationDetails?.ApplyData?.ApplyDetails?.ApplicationSubmittedOn != null)
+                model.ApplicationSubmittedOn = applicationDetails.ApplyData.ApplyDetails.ApplicationSubmittedOn;
+
+            var ukprn = await _qnaApiClient.GetQuestionTag(request.ApplicationId, "UKPRN");
+            model.Ukprn = ukprn;
+
+            var tradingNameAndWebsitePage = await _qnaApiClient.GetPageBySectionNo(request.ApplicationId, 0, 1, "1");
```
Comment (David-B-Read) [inline src/SFA.DAS.AdminService.Web/Handlers/Gateway/GetTradingNameHandler.cs]: Page, sequence and section ids should go into constants file rather using magic numbers

**Comment by martin-hornsby** [comment]
unit tests are failing too...

File: src/SFA.DAS.AdminService.Web/Views/Roatp/Apply/Gateway/Pages/TradingName.cshtml [hunk 13]
```diff
@@ -0,0 +1,307 @@
...
+            }
+
+            if (error.Field == "OptionFailText")
+            {
+                isErrorFailText = "govuk-textarea--error-override";
+                failTextMessage = error.ErrorMessage;
+            }
+        }
+    }
+    var applicationsListTab = "New";
```
Comment (David-B-Read) [inline src/SFA.DAS.AdminService.Web/Views/Roatp/Apply/Gateway/Pages/TradingName.cshtml]: Ideally this logic should be in C# code not in the view

**Pull request merged**